src/clustering/administration/servers/auto_reconnect.cc

```diff
@@ -84,10 +84,25 @@ void auto_reconnector_t::try_reconnect(const server_id_t &server,
 
     wait_any_t interruptor(&reconnected, &give_up_timer, keepalive.get_drain_signal());
 
+    cond_t join_failed;
     exponential_backoff_t backoff(50, 15 * 1000);
     try {
-        while (!interruptor.is_pulsed()) {
-            connectivity_cluster_run->join(last_known_address, join_delay_secs);
+        while (!interruptor.is_pulsed() && !join_failed.is_pulsed()) {
+            coro_t::spawn_now_dangerously([&]() {
```

Comment:
Actually there's another problem that the call to `try_reconnect` can finish while this coroutine is still running. Because `join_failed` is a local variable, and the coroutine accesses it, that would be illegal.

Instead of taking a copy of `keepalive`, I suggest instantiating a new `auto_drainer_t` locally in the `try_reconnect` method. Make sure you declare that drainer after `join_failed`, so that it is drained before `join_failed` gets destructed when `try_reconnect` returns. To make things more explicit, I suggest specifying all variable bindings explicitly in the lambda, rather than using the catch-all `&`:
```
auto_drainer_t::lock_t local_keepalive = local_drainer.lock();
coro_t::spawn_sometime(
    [local_keepalive /* must capture by value */, join_failed &, last_known_address, server, this]() {
```